src/librustc_infer/traits/wf.rs

```diff
@@ -323,8 +323,15 @@ impl<'a, 'tcx> WfPredicates<'a, 'tcx> {
             let implied_obligations = traits::elaborate_obligations(tcx, obligations.clone());
             let implied_obligations = implied_obligations.map(|obligation| {
                 let mut cause = cause.clone();
-                cause.code =
-                    traits::ObligationCauseCode::DerivedObligation(Rc::new(obligation.cause.code));
+                let parent_trait_ref = obligation
+                    .predicate
+                    .to_opt_poly_trait_ref()
+                    .unwrap_or_else(|| ty::Binder::dummy(*trait_ref));
```

Comment:
I think "parent" implies you should always use `trait_ref`, while `obligation.predicate` is used for the "child" obligation that this closure produces.

Reply:
I don't think that's right. Changing this to be `let parent_trait_ref = ty::Binder::dummy(*trait_ref);` causes the output to be nonsensical.

From:

```
error[E0277]: the trait bound `T: MyDisplay` is not satisfied
  --> $DIR/issue-65774-2.rs:16:6
   |
LL | trait MPU {
   |       ---
LL |     type MpuConfig: MyDisplay = T;
   |                     --------- required by this bound in `MPU`
...
LL | impl MPU for S { }
   |      ^^^ the trait `MyDisplay` is not implemented for `T`
   |
   = note: required because of the requirements on the impl of `MyDisplay` for `<S as MPU>::MpuConfig`
```
to
```
error[E0277]: the trait bound `T: MyDisplay` is not satisfied
  --> $DIR/issue-65774-2.rs:16:6
   |
LL | trait MPU {
   |       ---
LL |     type MpuConfig: MyDisplay = T;
   |                     --------- required by this bound in `MPU`
...
LL | impl MPU for S { }
   |      ^^^ the trait `MyDisplay` is not implemented for `T`
   |
   = note: required because of the requirements on the impl of `MPU` for `S`
```

In my eyes the first one is clearly correct, while the second one is clearly incorrect.

Reply:
How is that error even possible? Do we not check associated type defaults?!
cc @nikomatsakis did we create another type alias situation?

Reply:
The associated type default is malformed *inside the trait*.
There should be no errors in `impl`s, only one in the `trait` itself.

Reply:
One of the reasons I'm saying that is because this output:
```
   = note: required because of the requirements on the impl of `MPU` for `S`
```

Is *the only* meaningful cause for something that *the impl* caused, try examples that don't involve associated type defaults.

It only looks weird because *the whole error* is nonsensical to begin with.
It shouldn't error *in the `impl`* when the `trait` definition *is clearly wrong*.